sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java

```diff
@@ -174,83 +177,192 @@ private <T> Flux<T> wrapWithTracingIfEnabled(CosmosPagedFluxOptions pagedFluxOpt
 
     private Flux<FeedResponse<T>> byPage(CosmosPagedFluxOptions pagedFluxOptions, Context context) {
         AtomicReference<Instant> startTime = new AtomicReference<>();
+        AtomicLong feedResponseConsumerLatencyInNanos = new AtomicLong(0);
 
         Flux<FeedResponse<T>> result =
             wrapWithTracingIfEnabled(pagedFluxOptions, this.optionsFluxFunction.apply(pagedFluxOptions), context)
-            .doOnSubscribe(ignoredValue -> startTime.set(Instant.now()))
+            .doOnSubscribe(ignoredValue -> {
+                startTime.set(Instant.now());
+                feedResponseConsumerLatencyInNanos.set(0);
+            })
             .doOnEach(signal -> {
+
+                CosmosAsyncClient client = pagedFluxOptions.getCosmosAsyncClient();
+
+                boolean clientTelemetryEnabled = false;
+                boolean clientMetricsEnabled = false;
+                ConsistencyLevel consistencyLevel = ConsistencyLevel.EVENTUAL;
+
+                if (client != null) {
+                    ImplementationBridgeHelpers.CosmosAsyncClientHelper.CosmosAsyncClientAccessor clientAccessor =
+                        ImplementationBridgeHelpers
+                        .CosmosAsyncClientHelper
+                        .getCosmosAsyncClientAccessor();
+                    clientTelemetryEnabled = clientAccessor.isSendClientTelemetryToServiceEnabled(client);
+                    clientMetricsEnabled = clientAccessor.isClientTelemetryMetricsEnabled(client);
+                    consistencyLevel =
+                        BridgeInternal
+                            .getContextClient(pagedFluxOptions.getCosmosAsyncClient())
+                            .getConsistencyLevel();
+                }
```

Comment:
Performing all of the above computation on each next item will be very heavy. Why do we need to compute all of this on each signal. This will impact the performance. We should move all of this computation to outside of each next signal emitting function.

Reply:
Let's discuss - reason why it is in the doOnEach is that the cosmosClient in pagedFluxOptions is mutable - and is actually set when the tracing scope is defined (for each request) - also compared to making an I/O request I don't think the Compute overhead is significant here. I had this outside initially - but at least for the first iteration the CosmosClient wasn't set correctly (found it with unit test failing) - and given that the client is mutable in the PagedFlux I don't think this is safe.